include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.h

```diff
@@ -64,22 +65,118 @@ struct PortInfo {
 
-class InnerSymbolTable {
+namespace detail {
+LogicalResult verifyInnerRefs(Operation *op);
+} // namespace detail
 
+/// Table of inner_sym's
+class InnerSymbolTable {
 public:
   /// Return the name of the attribute used for inner symbol names.
   static StringRef getInnerSymbolAttrName() { return "inner_sym"; }
+
+  /// Build an inner_sym table for the given operation.
+  explicit InnerSymbolTable(Operation *op);
+
+  /// Look up a symbol with the specified name, returning null if no such
+  /// name exists. Names never include the @ on them.
+  Operation *lookup(StringRef name) const;
+  template <typename T>
+  T lookup(StringRef name) const {
+    return dyn_cast_or_null<T>(lookup(name));
+  }
+
+  /// Look up a symbol with the specified name, returning null if no such
+  /// name exists. Names never include the @ on them.
+  Operation *lookup(StringAttr name) const;
+  template <typename T>
+  T lookup(StringAttr name) const {
+    return dyn_cast_or_null<T>(lookup(name));
+  }
+
+private:
+  /// This is the operation this table is constructed for.
+  /// Must have InnerSymbolTable trait.
+  Operation *innerSymTblOp;
+
+  /// This maps names to operations with that inner symbol.
+  DenseMap<StringAttr, Operation *> symbolTable;
+};
+
+/// Collection of InnerSymbolTables and utilities for constructing them
+class InnerSymbolTableCollection {
+public:
+  /// Get or create the InnerSymbolTable for the specified operation
+  InnerSymbolTable &getInnerSymbolTable(Operation *op);
+
+  /// Ensure the specified operations have tables, creating them in parallel.
+  /// The operation list must not contain duplicates.
+  void constructTablesInParallelFor(ArrayRef<Operation *> ops);
+
+private:
+  /// This maps Operations to their InnnerSymbolTables, constructed lazily
+  DenseMap<Operation *, std::unique_ptr<InnerSymbolTable>> symbolTables;
+};
+
+/// Represents the namespace in which InnerRef's can be resolved.
+struct InnerRefNamespace {
+  SymbolTable &symTable;
+  InnerSymbolTableCollection &innerSymTables;
+
+  /// Resolve the InnerRef to the operation it targets.
+  /// Returns null if no such name exists.
+  /// Note that some InnerRef's target ports and must be handled separately.
+  Operation *lookup(hw::InnerRefAttr inner);
+  template <typename T>
+  T lookup(hw::InnerRefAttr inner) {
+    return dyn_cast_or_null<T>(lookup(inner));
+  }
 };
 
 } // namespace firrtl
 } // namespace circt
 
 namespace mlir {
 namespace OpTrait {
+
+/// Scope for resolving innerref's
+/// Must also be a SymbolTable.
+/// Immediate children with the InnerSymbolTable trait will be resolved against.
+template <typename ConcreteType>
+class InnerRefNamespace : public TraitBase<ConcreteType, InnerRefNamespace> {
+public:
+  static LogicalResult verifyRegionTrait(Operation *op) {
+    if (!op->hasTrait<OpTrait::SymbolTable>())
+      return op->emitError(
+          "InnerRefNamespace must also have SymbolTable trait");
```

Comment:
Can this be a static assert? There is some precedence for this. e.g. [here]( https://github.com/llvm/circt/blob/main/include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td#L258).  Same comment below.

Reply:
Oh, much better!  Was bugging me, thanks!